app/src/main/java/io/appium/settings/LocationService.java

```diff
@@ -172,45 +170,61 @@ public void run() {
         locationUpdatesTimer.schedule(locationUpdateTask, 0, UPDATE_INTERVAL_MS);
     }
 
-    private void updateMockLocationFactory(Intent intent) {
+    private Location getLocation(Intent intent, String providerName) {
         double longitude;
         try {
-            longitude = Double.valueOf(intent.getStringExtra("longitude"));
+            longitude = Double.parseDouble(intent.getStringExtra("longitude"));
         } catch (NumberFormatException e) {
             throw new IllegalArgumentException(
                     String.format("longitude should be a valid number. '%s' is given instead",
                             intent.getStringExtra(LONGITUDE_PARAMETER_KEY)));
         }
         double latitude;
         try {
-            latitude = Double.valueOf(intent.getStringExtra("latitude"));
+            latitude = Double.parseDouble(intent.getStringExtra("latitude"));
         } catch (NumberFormatException e) {
             throw new IllegalArgumentException(
                     String.format("latitude should be a valid number. '%s' is given instead",
                             intent.getStringExtra(LATITUDE_PARAMETER_KEY)));
         }
-        double altitude = 0.0;
+
+        double altitude = 0.0f;
         try {
             if (intent.hasExtra(ALTITUDE_PARAMETER_KEY)) {
-                altitude = Double.valueOf(intent.getStringExtra(ALTITUDE_PARAMETER_KEY));
+                altitude = Double.parseDouble(intent.getStringExtra(ALTITUDE_PARAMETER_KEY));
             }
         } catch (NumberFormatException e) {
             Log.e(TAG, String.format("altitude should be a valid number. '%s' is given instead",
                     intent.getStringExtra(ALTITUDE_PARAMETER_KEY)));
         }
+
+        float speed = 0.0f;
         try {
```

Comment:
I tried, but I found there're not many benefits, because it has 2 return types `double` and `float`, as well as the 2 different behavours `optional` & `required` value, even if we have that helper method, we have to pass too many args.

Reply:
for double/float we could simply use double as it is a superclass of float anyway.
Regarding optional/required I'm ok to have one additional argument. Basically, it is not mandatory to have everything as arguments, but just have one represented as a data object

Reply:
I found out the if the `required` param `longitude` and `latitude` is not passed, the service will directly crash due to the exceptions we thrown.
We should not crash the service, so there's no required param anymore, just log some warnings.

what do you think?

Reply:
yes, this would make sense. We might simply write an error to the log without crashing it